File: examples/poll_pv.py
"""Poll a PV on a fixed interval and stream the values.

No events or polling configuration required on the IOC side.

Usage:
    python poll_pv.py <pv_name> [interval-ms]

Examples:
    python poll_pv.py TEST:CALC
    python poll_pv.py TEST:CALC 500
"""

import asyncio
import sys
import time
from datetime import timedelta
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

import reactivex as rx
import reactivex.operators as ops
from reactivex.scheduler.eventloop import AsyncIOScheduler
from caproto.asyncio.client import Context

from rxepics.channel import read_pv


async def main():
    if len(sys.argv) < 2:
        print("Usage: poll_pv.py <pv_name> [interval-ms]", file=sys.stderr)
        sys.exit(1)

    pv_name = sys.argv[1]
    interval_ms = int(sys.argv[2]) if len(sys.argv) > 2 else 1000

    loop = asyncio.get_running_loop()
    scheduler = AsyncIOScheduler(loop)
    ctx = Context()

    print(f"Polling {pv_name} every {interval_ms} ms — Ctrl+C to stop")

    # interval() ticks every N ms and triggers a fresh read on each tick.
    # flat_map turns each tick into a one-shot read Observable.
    # No loop. No thread management. No counter.
    rx.interval(timedelta(milliseconds=interval_ms), scheduler=scheduler).pipe(
        ops.flat_map(lambda _: read_pv(pv_name, ctx))
    ).subscribe(
        on_next=lambda v: print(f"[{int(time.time() * 1000)}]  {v}"),
        on_error=lambda e: print(f"ERROR: {e}", file=sys.stderr),
        scheduler=scheduler,
    )

    await asyncio.Future()  # run until Ctrl+C


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        pass

File: RxEpics/cpp/examples/poll_pv.cpp
/**
 * Poll an EPICS PV at a fixed interval — no loop, no thread.
 *
 * Mirrors Python's poll_pv.py. Prefer monitor_pv for IOCs with scan records;
 * use this when you need client-side rate control independent of the IOC scan.
 *
 * Usage:
 *   EPICS_PVA_ADDR_LIST=localhost ./poll_pv [pv] [interval-ms]
 *   defaults: TEST:CALC  500
 */

#include <chrono>
#include <csignal>
#include <iostream>
#include <string>

#include <rxcpp/rx.hpp>
#include <rxepics/rxepics.hpp>

static volatile sig_atomic_t g_running = 1;

int main(int argc, char* argv[]) {
    const std::string pv          = argc > 1 ? argv[1] : "TEST:CALC";
    const int         interval_ms = argc > 2 ? std::stoi(argv[2]) : 500;

    std::signal(SIGINT, [](int) { g_running = 0; });

    auto& ctx = rxepics::default_context();

    std::cout << "Polling " << pv << " every " << interval_ms << " ms  (Ctrl+C to stop)\n\n";

    auto sub = rxcpp::observable<>::interval(std::chrono::milliseconds(interval_ms))
        .flat_map([pv, &ctx](long) { return rxepics::read_pv<double>(pv, ctx); })
        .subscribe(
            [](double v) { std::cout << "  " << v << "\n"; },
            [](std::exception_ptr e) {
                try { std::rethrow_exception(e); }
                catch (std::exception& ex) { std::cerr << "  ERROR: " << ex.what() << "\n"; }
            }
        );

    while (g_running) std::this_thread::sleep_for(std::chrono::milliseconds(100));
    sub.unsubscribe();
    std::cout << "\n  stopped.\n";
    return 0;
}
